src/cm/communication_manager.rs

```diff
@@ -56,55 +229,356 @@ impl EventChannel {
-        let mut cm_id = MaybeUninit::<*mut rdma_cm_id>::uninit();
-        let ret;
-
-        unsafe {
-            ret = rdma_create_id(
-                self.channel.as_mut(),
-                cm_id.as_mut_ptr(),
-                ctx.context as _,
-                PortSpace::Tcp as u32,
-            );
-        }
+    pub fn create_id(&mut self, port_space: PortSpace) -> Result<Arc<Identifier>, String> {
+        let mut cm_id_ptr: *mut rdma_cm_id = null_mut();
+        let ret = unsafe { rdma_create_id(self.channel.as_mut(), &mut cm_id_ptr, null_mut(), port_space as u32) };
 
         if ret < 0 {
             return Err(format!("Failed to create cm_id {ret}"));
         }
 
-        unsafe {
-            cm_id.assume_init();
+        new_cm_id_for_raw(cm_id_ptr)
+    }
+
+    pub fn get_cm_event(&mut self) -> Result<Event, String> {
+        let mut event_ptr = MaybeUninit::<*mut rdma_cm_event>::uninit();
+
+        let ret = unsafe { rdma_get_cm_event(self.channel.as_ptr(), event_ptr.as_mut_ptr()) };
+
+        if ret < 0 {
+            return Err(format!("Failed to get cm event {:?}", io::Error::last_os_error()));
         }
 
-        Ok(CommunicationManager {
-            cm_id: unsafe { NonNull::new(*cm_id.as_mut_ptr()).unwrap_unchecked() },
+        let event = unsafe { NonNull::new(event_ptr.assume_init()).unwrap() };
+
+        let cm_id = unsafe {
+            let raw_cm_id = event.as_ref().id;
+
+            assert_ne!(raw_cm_id, null_mut());
+            if event.as_ref().event == EventType::ConnectRequest as u32 {
+                // For connect requests, create a new CommunicationManager
+                Some(new_cm_id_for_raw(raw_cm_id).unwrap())
+            } else {
+                // For other events, return the existing CommunicationManager
+                let context_ptr = (*raw_cm_id).context as *mut Weak<Identifier>;
+                assert_ne!(context_ptr, null_mut());
+                (*context_ptr).clone().upgrade()
+            }
+        };
+
+        let listener_id = unsafe {
+            let raw_listen_id = event.as_ref().listen_id;
+
+            if !raw_listen_id.is_null() {
+                let context_ptr = (*raw_listen_id).context as *mut Weak<Identifier>;
+                assert_ne!(context_ptr, null_mut());
+                (*context_ptr).clone().upgrade()
+            } else {
+                None
+            }
+        };
+
+        Ok(Event {
+            event,
+            cm_id,
+            listener_id,
         })
     }
+}
+
+unsafe impl Send for EventChannel {}
+unsafe impl Sync for EventChannel {}
 
-    pub fn get_cm_event() -> Result<Event, String> {
-        todo!();
+impl Drop for Identifier {
+    fn drop(&mut self) {
+        let cm_id = self.cm_id;
+        unsafe {
+            let _ = Box::from_raw((*cm_id.as_ptr()).context as *mut Weak<Identifier>);
+            rdma_destroy_id(cm_id.as_ptr());
```

Comment:
This order is still problematic since `get_cm_event()` may dereference a wild pointer to `Weak<Identifier>`, which is already destructed before `rdma_destroy_id()` is invoked.
Consider using the following code instead:
```Rust
        unsafe {
            let ctx = cm_id.as_ref().context as *mut Weak<Identifier>;
            rdma_destroy_id(cm_id.as_ptr());
            let _ = Box::from_raw(ctx);
        }
```

Reply:
You are absolutely right 👍